File: behavioral-design-patterns/observer-design-pattern/Typescript/order-status.ts
// Abstract Observer
class Observer {
    update(order: Order) {}
}

// Concrete Observers
class Customer extends Observer {
    private name: string;

    constructor(name: string) {
        super();
        this.name = name;
    }

    update(order: Order){
        console.log(`Hello ${this.name}! Order #${order.getId()} is now ${order.getStatus()}.`);
    }
}

class DeliveryPartner extends Observer {
    private name: string;

    constructor(name: string) {
        super();
        this.name = name;
    }

    update(order: Order){
        console.log(`Driver ${this.name}: Order #${order.getId()} is now ${order.getStatus()}.`);
    }
}
class Restaurant extends Observer {
    private name: string;

    constructor(name: string) {
        super();
        this.name = name;
    }

    update(order: Order){
        console.log(`Restaurant ${this.name}: Order #${order.getId()} is now ${order.getStatus()}.`);
    }
}
class CallCenter extends Observer {
    constructor() {
        super();
    }

    update(order: Order){
        console.log(`Call Center: Order #${order.getId()} is now ${order.getStatus()}.`);
    }
}

// Subject
class Order {

    private id: number;
    private status: string;
    private observers: Observer[] = [];

    constructor(id: number) {
        this.id = id;
        this.status = "Order Placed";
    }

    public getId(){
        return this.id;
    }

    public getStatus(){
        return this.status;
    }

    public setStatus(status: string){
        this.status = status;
        this.notifyObservers();
    }

    public attach(observer: Observer){
        this.observers.push(observer);
    }

    public detach(observer: Observer){
        this.observers = this.observers.filter(item => item !== observer);
    }

    private notifyObservers(){
        for (const observer of this.observers) {
            observer.update(this);
        }
    }
}

// Client
function notificationManager() {
    // Create an order
    const order1: Order = new Order(123);

    // Create customers, restaurants, drivers, and a call center to track the order
    const customer: Customer = new Customer("CustomerA");
    const driver: DeliveryPartner = new DeliveryPartner("DriverA");
    const restaurant: Restaurant = new Restaurant("RestaurantA");
    const callCenter: CallCenter = new CallCenter();

    // Attach observers to the order
    order1.attach(customer);
    order1.attach(driver);
    order1.attach(restaurant);
    order1.attach(callCenter);

    // Simulate order status updates
    order1.setStatus("out for delivery");

    console.log("");
    
    // Detach an observer (if needed)
    order1.detach(callCenter);

    // Simulate more order status updates
    order1.setStatus("delivered");
}

notificationManager()

File: behavioral-design-patterns/observer-design-pattern/Python/order-status.py
from typing import List

# Abstract Observer
class Observer:
    def __init__(self) -> None:
        pass
    
    def update(self, order):
        pass
    
# Concrete Observers
class Customer(Observer):
    def __init__(self, name) -> None:
        super().__init__()
        self.name = name
        
    def update(self, order):
        print(f"Hello, {self.name}! Order # {order.get_id()} is now {order.get_status()}.")

class DeliveryPartner(Observer):
    def __init__(self, driver_name) -> None:
        super().__init__()
        self.driver_name = driver_name
        
    def update(self, order):
        print(f"Driver {self.driver_name}: Order # {order.get_id()} is now {order.get_status()}.")

class Restaurant(Observer):
    def __init__(self, restaurant_name) -> None:
        super().__init__()
        self.restaurant_name = restaurant_name
        
    def update(self, order):
        print(f"Restaurant {self.restaurant_name}: Order # {order.get_id()} is now {order.get_status()}.")

class CallCenter(Observer):
    def __init__(self) -> None:
        super().__init__()
        
    def update(self, order):
        print(f"Call center: Order # {order.get_id()} is now {order.get_status()}.")
        
# Subject
class Order:    
    def __init__(self, id) -> None:
        self.__id = id
        self.__status = "Order Placed"
        self.__observers: List[Observer] = []
        
    def get_id(self):
        return self.__id
    
    def get_status(self):
        return self.__status
    
    def set_status(self, status_message: str):
        self.__status = status_message
        self.notify_observers()
        
    def attach(self, observer: Observer):
        self.__observers.append((observer))
        
    def detach(self, observer: Observer):
        self.__observers.remove(observer)
        
    def notify_observers(self):
        for observer in self.__observers:
            observer.update(self)
            
# Client
def main():
    # Create an order
    order1: Order = Order(id=123)
    
    # Create customers, restaurants, drivers, and a call center to track the order
    customer: Customer = Customer(name="CustomerA")
    restaurant: Restaurant = Restaurant(restaurant_name="RestaurantA")
    driver: DeliveryPartner = DeliveryPartner(driver_name="DriverC")
    call_center: CallCenter = CallCenter()
    
    # Attach observers to the order
    order1.attach(customer)
    order1.attach(restaurant)
    order1.attach(driver)
    order1.attach(call_center)
    
    # Simulate order status updates
    order1.set_status(status_message="out for delivery")
    
    print()
    
    # Detach an observer (if needed)
    order1.detach(call_center)
    
    # Simulate more order status updates
    order1.set_status(status_message="delivered")
    
# calling main method
if __name__ == "__main__":
    main()